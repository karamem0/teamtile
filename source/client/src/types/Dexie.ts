//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface ArrayEntity<T> {
  id: string,
  expired: number,
  values: T[]
}

export interface ValueEntity<T> {
  id: string,
  expired: number,
  value: T
}
