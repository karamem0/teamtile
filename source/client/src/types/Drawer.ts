//
// Copyright (c) 2021-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export type DrawerType =
  | 'channel'
  | 'member'
  | 'tag';
