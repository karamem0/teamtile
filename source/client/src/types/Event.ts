//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

export type Event = React.SyntheticEvent | React.SyntheticEvent<unknown> | Record<string, never>;

export type EventHandler<T = never> = (
  event?: Event,
  data?: T
) => void;
