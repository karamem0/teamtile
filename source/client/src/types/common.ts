//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface KeyValue<K, V> {
  key: K,
  value: V
}
