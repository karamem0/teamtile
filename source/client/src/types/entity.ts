//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface Channel {
  displayName: string | null | undefined,
  id: string | undefined,
  membershipType: MembershipType | null | undefined,
  webUrl: string | null | undefined
}

export interface Group {
  id: string | undefined,
  mail: string | null | undefined,
  sensitivityLabel: string | null | undefined
}

export interface Drive {
  id: string | undefined,
  webUrl: string | null | undefined
}

export interface Icon {
  data: string | null | undefined,
  type: string | null | undefined
}

export interface Member {
  displayName: string | null | undefined,
  email: string | null | undefined,
  id: string | undefined,
  userId: string | null | undefined
}

export type MemberWithIcon = (
  Member & {
     icon: Icon | undefined
  }
)

export interface Tab {
  appId: string | null | undefined,
  displayName: string | null | undefined,
  id: string | undefined,
  webUrl: string | null | undefined
}

export interface Team {
  description: string | null | undefined,
  displayName: string | null | undefined,
  id: string | undefined,
  internalId: string | null | undefined,
  visibility: VisibilityType | null | undefined,
  webUrl: string | null | undefined
}

export type TeamWithMail = (
  Team & {
     mail: string | null | undefined
  }
)

export type TeamWithIcon = (
  Team & {
     icon: Icon | undefined
  }
)

export enum MembershipType {
  private = 'private',
  standard = 'standard'
}

export enum VisibilityType {
  private = 'private',
  public = 'public'
}
