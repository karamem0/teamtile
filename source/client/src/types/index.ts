//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface Team {
  id?: string;
  internalId?: string;
  name?: string;
  description?: string;
  visibility?: string;
  url?: string;
  icon?: string;
}

export interface TeamChannels {
  count?: number;
  nextLink?: string;
  values?: TeamChannel[];
}

export interface TeamChannel {
  id?: string;
  name?: string;
  url?: string;
}

export interface TeamMembers {
  count?: number;
  nextLink?: string;
  values?: TeamMember[];
}

export interface TeamMember {
  id?: string;
  name?: string;
  email?: string;
}

export interface TeamDrive {
  id?: string;
  url?: string;
}
