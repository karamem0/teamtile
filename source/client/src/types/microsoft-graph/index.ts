import * as microsoftGraph from '@microsoft/microsoft-graph-types';

export type MicrosoftGraphResponse = MicrosoftGraphValueResponse | MicrosoftGraphArrayResponse | MicrosoftGraphErrorResponse | string;

export interface MicrosoftGraphValueResponse extends microsoftGraph.Entity
{
  '@odata.context': string;
}

export interface MicrosoftGraphArrayResponse {
  '@odata.context': string;
  '@odata.count': number;
  '@odata.nextLink': string;
  value: microsoftGraph.Entity[];
}

export interface MicrosoftGraphBatchResponse {
  id: string;
  status: number;
  body: MicrosoftGraphResponse;
  headers: {
    [key: string]: string;
  };
}

export interface MicrosoftGraphErrorResponse {
  error: microsoftGraph.GenericError;
}
