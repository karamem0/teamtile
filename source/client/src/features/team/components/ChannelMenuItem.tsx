//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';
import {
  useAsyncFn,
  useDebounce,
  useError
} from 'react-use';

import { InputProps } from '@fluentui/react-northstar';

import { app } from '@microsoft/teams-js';

import { Channel } from '../../../types/Entity';
import { Item } from '../../../types/Store';
import { search } from '../../../utils/String';
import { getChannels } from '../managers/TeamManager';

import Presenter from './ChannelMenuItem.presenter';

interface ChannelMenuItemProps {
  item?: Item
}

function ChannelMenuItem(props: ChannelMenuItemProps) {

  const { item } = props;

  const dispatchError = useError();
  const [ state, fetch ] = useAsyncFn((teamId: string) => getChannels(teamId));

  const [ values, setValues ] = React.useState<Channel[]>();
  const [ filter, setFilter ] = React.useState<string>();

  React.useEffect(() => {
    if (!state.error) {
      return;
    }
    dispatchError(state.error);
  }, [
    dispatchError,
    state.error
  ]);

  useDebounce(() => {
    setValues(state.value?.filter((item) => search(item.displayName, filter)));
  }, 500, [
    state.value,
    filter
  ]);

  const handleClick = React.useCallback((_, data?: Channel) => {
    if (!data?.webUrl) {
      return;
    }
    app.openLink(data.webUrl);
  }, []);

  const handleFilterChange = React.useCallback((_, data?: InputProps & { value: string }) => {
    setFilter(data?.value);
  }, []);

  const handleOpenChange = React.useCallback(async (_, data?: boolean) => {
    if (!data) {
      return;
    }
    if (!item?.id) {
      return;
    }
    await fetch(item.id);
  }, [
    fetch,
    item
  ]);

  return (
    <Presenter
      items={values}
      loading={state.loading}
      onClick={handleClick}
      onFilterChange={handleFilterChange}
      onOpenChange={handleOpenChange} />
  );

}

export default ChannelMenuItem;
