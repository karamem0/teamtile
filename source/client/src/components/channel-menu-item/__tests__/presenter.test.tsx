//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';

import { Channel } from '../../../types/entity';
import ChannelMenuItem from '../presenter';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('ChannelMenuItem', () => {

  it('create shapshot when menu item is not open', async () => {
    const json = await import('./__jsons__/presenter.test.json');
    const params = {
      channels: json.default as Channel[],
      onClick: jest.fn()
    };
    render(<ChannelMenuItem {...params} />);
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

  it('create shapshot when menu item is open', async () => {
    const json = await import('./__jsons__/presenter.test.json');
    const params = {
      channels: json.default as Channel[],
      onClick: jest.fn()
    };
    render(<ChannelMenuItem {...params} />);
    fireEvent.click(screen.getByRole('button'));
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
