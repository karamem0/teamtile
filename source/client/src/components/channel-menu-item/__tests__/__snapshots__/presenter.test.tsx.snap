// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelMenuItem create shapshot when menu item is not open 1`] = `
.emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  cursor: pointer;
}

.emotion-1 {
  line-height: 1rem;
}

.emotion-2 {
  width: 1rem;
  height: 1rem;
}

<body>
  <div>
    <span
      class="ui-text emotion-0"
      data-uses-unhanded-props="true"
      role="button"
    >
      <span
        class="emotion-1"
      >
        <span
          aria-hidden="true"
          class="root_dd790ee3 emotion-2"
        >
          <svg
            class="svg_dd790ee3"
            focusable="false"
            viewBox="0 0 2048 2048"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1792 128v1664H256V128h1536zm-128 128H384v1408h1280V256zm-256 768H640V896h768v128zm0 384H640v-128h768v128zm0-768H640V512h768v128z"
            />
          </svg>
        </span>
      </span>
      <span
        class="ui-text emotion-1"
        data-uses-unhanded-props="true"
      >
        3
      </span>
    </span>
  </div>
</body>
`;

exports[`ChannelMenuItem create shapshot when menu item is open 1`] = `
.emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  cursor: pointer;
}

.emotion-1 {
  line-height: 1rem;
}

.emotion-2 {
  width: 1rem;
  height: 1rem;
}

.emotion-4 {
  width: 12rem;
  max-height: 24rem;
  margin: -0.5rem;
  overflow: auto;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: column;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  gap: 0.25rem;
}

.emotion-6 {
  display: grid;
  grid-template-columns: auto auto;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -ms-flex-pack: left;
  -webkit-justify-content: left;
  justify-content: left;
  margin: 0 -0.5rem;
}

<body>
  <div>
    <span
      class="ui-text emotion-0"
      data-uses-unhanded-props="true"
      role="button"
    >
      <span
        class="emotion-1"
      >
        <span
          aria-hidden="true"
          class="root_dd790ee3 emotion-2"
        >
          <svg
            class="svg_dd790ee3"
            focusable="false"
            viewBox="0 0 2048 2048"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1792 128v1664H256V128h1536zm-128 128H384v1408h1280V256zm-256 768H640V896h768v128zm0 384H640v-128h768v128zm0-768H640V512h768v128z"
            />
          </svg>
        </span>
      </span>
      <span
        class="ui-text emotion-1"
        data-uses-unhanded-props="true"
      >
        3
      </span>
    </span>
  </div>
  <div
    class="ui-popup__content ui-animation"
    data-aa-class="Popup__popup"
    data-uses-unhanded-props="true"
    role="complementary"
    style="position: fixed; left: 0px; top: 0px; margin: 0px;"
  >
    <div
      class="ui-popup__content__content"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        >
          <div
            class="ui-box ui-input"
            data-uses-unhanded-props="true"
          >
            <div
              class="ui-box"
              data-uses-unhanded-props="true"
            >
              <input
                class="ui-box ui-input__input"
                data-aa-class="Input__input"
                data-uses-unhanded-props="true"
                id="ui-input-1"
                type="text"
                value=""
              />
              <div
                class="ui-box ui-input__icon"
                data-uses-unhanded-props="true"
              >
                <span
                  aria-hidden="true"
                  class="root_dd790ee3"
                >
                  <svg
                    class="svg_dd790ee3"
                    focusable="false"
                    viewBox="0 0 2048 2048"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M1344 0q97 0 187 25t168 71 142 110 111 143 71 168 25 187q0 97-25 187t-71 168-110 142-143 111-168 71-187 25q-125 0-239-42t-211-121l-785 784q-19 19-45 19t-45-19-19-45q0-26 19-45l784-785q-79-96-121-210t-42-240q0-97 25-187t71-168 110-142T989 96t168-71 187-25zm0 1280q119 0 224-45t183-124 123-183 46-224q0-119-45-224t-124-183-183-123-224-46q-119 0-224 45T937 297 814 480t-46 224q0 119 45 224t124 183 183 123 224 46z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <ul
            class="ms-FocusZone ui-list"
            data-aa-class="List"
            data-focuszone-id="FocusZone2"
            data-uses-unhanded-props="true"
            role="menu"
          >
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="0"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-6"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        General
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-6"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        Onboarding
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-6"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        Benefits
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
